src/main/java/org/jabref/logic/importer/fetcher/IEEE.java

```diff
@@ -224,12 +226,30 @@ public String getName() {
         return Optional.of(HelpFile.FETCHER_IEEEXPLORE);
     }
 
+    /**
+     * Gets IEEE api key, use key if it customized it in the preferences, otherwise use the default
+     *
+     * @return IEEE API Key
+     */
+    private String getApiKey() {
+        // TODO: Construct CustomPreferences as a global variable, and may modify ImportFormatPreferences
+        JabRefPreferences preferences = null;
+        String apiKey = API_KEY;
+        if (preferences != null) {
+            CustomApiKeyPreferences apiKeyPreferences = preferences.getCustomApiKeyPreferences(getName());
+            if (apiKeyPreferences.isUseCustom()) {
+                apiKey = apiKeyPreferences.getDefaultApiKey();
+            }
+        }
+        return apiKey;
+    }
```

Comment:
?

Reply:
I'm sorry to have taken an incorrect name here, defaultApiKey should be renamed to customApiKey.

```
public CustomApiKeyPreferences getCustomApiKeyPreferences(String name) {
    return new CustomApiKeyPreferences(name,
            getBoolean(USE_CUSTOM_API_KEY + name, false),
            get(CUSTOM_API_KEY + name, ""));
}
```
If the API Key is not saved in the preferences, getCustomApiKeyPreferences just returns a default value, in fact, the default value is not saved in the preferences.

One thing that can be modified is that if the API Key requested to save is empty, then clear it

Reply:
Yes, of course it should store the custom key in the preferences
The preferences and the "default preferences" should not have the default key, which is provided by the build info/GitHub secret.

Reply:
The default values are not saved in the preferences.

One question now is how to pass PreferencesService into Fetcher, how about modifying the input parameter ImportFormatPreferences of the getSearchBasedFetchers method to PreferencesService?

Reply:
> The default values are not saved in the preferences.

So what is `private String defaultApiKey` for? The custom key?

Please get your wording straight. It is impossible to discuss your PR if we continuously talk about different things because your variable names are misleading.

> One question now is how to pass PreferencesService into Fetcher, how about modifying the input parameter ImportFormatPreferences of the getSearchBasedFetchers method to PreferencesService?

This is exactly what I said earlier. https://github.com/JabRef/jabref/pull/7720#issuecomment-835856203 https://github.com/JabRef/jabref/pull/7720#issuecomment-842875913
But don't modify ImportFormatPreferences - replace it with PreferencesService.

Reply:
I am modifying the variable name, but I want to do PreferencesService related things at the same time.